File: backend/routers/articles.py
from fastapi import APIRouter, HTTPException
from sqlalchemy import select
from backend.database import AsyncSessionLocal
from backend.models_db import Article, ArticleVersion
from backend.models import ArticlePatchRequest

router = APIRouter(prefix="/articles")

@router.get("")
async def list_articles():
    """返回文章列表（不含全文），按 created_at 降序"""
    async with AsyncSessionLocal() as session:
        result = await session.execute(
            select(Article).order_by(Article.created_at.desc())
        )
        articles = result.scalars().all()
    return [
        {
            "id": a.id,
            "job_id": a.job_id,
            "topic": a.topic,
            "word_count": a.word_count,
            "created_at": a.created_at.isoformat(),
        }
        for a in articles
    ]

@router.get("/{article_id}")
async def get_article(article_id: str):
    """返回单篇文章全文"""
    async with AsyncSessionLocal() as session:
        article = await session.get(Article, article_id)
    if not article:
        raise HTTPException(status_code=404, detail="Article not found")
    return {
        "id": article.id,
        "job_id": article.job_id,
        "topic": article.topic,
        "content": article.content,
        "word_count": article.word_count,
        "created_at": article.created_at.isoformat(),
    }


@router.patch("/{article_id}")
async def patch_article(article_id: str, req: ArticlePatchRequest):
    """更新文章内容，同时追加一条历史快照"""
    async with AsyncSessionLocal() as session:
        article = await session.get(Article, article_id)
        if not article:
            raise HTTPException(status_code=404, detail="Article not found")
        # 先把当前内容存为历史快照，再覆盖，防止原始版本丢失
        old_version = ArticleVersion(
            article_id=article_id,
            content=article.content,
            word_count=article.word_count,
        )
        session.add(old_version)
        article.content = req.content
        article.word_count = len(req.content.replace(" ", ""))
        await session.commit()
    return {"status": "ok"}


@router.get("/{article_id}/versions")
async def list_versions(article_id: str):
    """返回该文章所有历史快照（不含全文），按时间降序"""
    async with AsyncSessionLocal() as session:
        article = await session.get(Article, article_id)
        if not article:
            raise HTTPException(status_code=404, detail="Article not found")
        result = await session.execute(
            select(ArticleVersion)
            .where(ArticleVersion.article_id == article_id)
            .order_by(ArticleVersion.saved_at.desc())
        )
        versions = result.scalars().all()
    return {
        "versions": [
            {"id": v.id, "saved_at": v.saved_at.isoformat(), "word_count": v.word_count}
            for v in versions
        ]
    }


@router.get("/{article_id}/versions/{version_id}")
async def get_version(article_id: str, version_id: int):
    """返回某条历史快照的完整内容"""
    async with AsyncSessionLocal() as session:
        version = await session.get(ArticleVersion, version_id)
        if not version or version.article_id != article_id:
            raise HTTPException(status_code=404, detail="Version not found")
        return {"id": version.id, "content": version.content, "saved_at": version.saved_at.isoformat()}


@router.post("/{article_id}/versions/{version_id}/restore")
async def restore_version(article_id: str, version_id: int):
    """将某条历史恢复为当前版本（等同于一次 PATCH）"""
    async with AsyncSessionLocal() as session:
        version = await session.get(ArticleVersion, version_id)
        if not version or version.article_id != article_id:
            raise HTTPException(status_code=404, detail="Version not found")
        article = await session.get(Article, article_id)
        if not article:
            raise HTTPException(status_code=404, detail="Article not found")
        article.content = version.content
        article.word_count = version.word_count
        new_version = ArticleVersion(
            article_id=article_id,
            content=version.content,
            word_count=version.word_count,
        )
        session.add(new_version)
        await session.commit()
    return {"status": "ok"}

File: packages/db/src/repositories/articles.ts
import { createHash } from 'node:crypto'
import { and, desc, eq, sql } from 'drizzle-orm'
import type { PgQueryResultHKT } from 'drizzle-orm/pg-core'
import {
  articles,
  articleVersions,
  jobs,
  type ArticleRow,
  type ArticleVersionRow,
} from '../schema'
import type { VibeDatabase } from './jobs'

const MAX_MARKDOWN_BYTES = 8 * 1024 * 1024

export type ArticleWriteInput = {
  articleId: string
  content: string
  expectedRevision: number
}

export type RestoreArticleVersionInput = {
  articleId: string
  versionId: number
  expectedRevision: number
}

export type ArticleMutationResult =
  | { status: 'updated'; article: ArticleRow; snapshot: ArticleVersionRow }
  | { status: 'not_found' }
  | { status: 'version_not_found' }
  | { status: 'revision_conflict'; currentRevision: number }

function normalizeContent(content: string): string {
  if (!content.trim()) throw new Error('Article content is required')
  if (Buffer.byteLength(content, 'utf8') > MAX_MARKDOWN_BYTES) {
    throw new Error(`Article content exceeds ${MAX_MARKDOWN_BYTES} bytes`)
  }
  return content
}

function contentFingerprint(content: string): string {
  return `sha256:${createHash('sha256').update(content).digest('hex')}`
}

function wordCount(content: string): number {
  return Array.from(content.replace(/\s/gu, '')).length
}

export class ArticleRepository<TQueryResult extends PgQueryResultHKT> {
  constructor(private readonly db: VibeDatabase<TQueryResult>) {}

  listArticles() {
    return this.db.select().from(articles).orderBy(desc(articles.createdAt))
  }

  listArticlesForWorkspace(workspaceId: string) {
    return this.db
      .select({ article: articles })
      .from(articles)
      .innerJoin(jobs, eq(jobs.id, articles.jobId))
      .where(eq(jobs.workspaceId, workspaceId))
      .orderBy(desc(articles.createdAt))
      .then((rows) => rows.map((row) => row.article))
  }

  async getArticle(articleId: string): Promise<ArticleRow | null> {
    const [article] = await this.db
      .select()
      .from(articles)
      .where(eq(articles.id, articleId))
      .limit(1)
    return article ?? null
  }

  async getArticleForWorkspace(
    articleId: string,
    workspaceId: string,
  ): Promise<ArticleRow | null> {
    const [row] = await this.db
      .select({ article: articles })
      .from(articles)
      .innerJoin(jobs, eq(jobs.id, articles.jobId))
      .where(and(eq(articles.id, articleId), eq(jobs.workspaceId, workspaceId)))
      .limit(1)
    return row?.article ?? null
  }

  async listVersions(articleId: string): Promise<ArticleVersionRow[] | null> {
    const [article] = await this.db
      .select({ id: articles.id })
      .from(articles)
      .where(eq(articles.id, articleId))
      .limit(1)
    if (!article) return null
    return this.db
      .select()
      .from(articleVersions)
      .where(eq(articleVersions.articleId, articleId))
      .orderBy(desc(articleVersions.savedAt), desc(articleVersions.id))
  }

  async listVersionsForWorkspace(
    articleId: string,
    workspaceId: string,
  ): Promise<ArticleVersionRow[] | null> {
    if (!(await this.getArticleForWorkspace(articleId, workspaceId))) return null
    return this.db
      .select()
      .from(articleVersions)
      .where(eq(articleVersions.articleId, articleId))
      .orderBy(desc(articleVersions.savedAt), desc(articleVersions.id))
  }

  async getVersion(articleId: string, versionId: number): Promise<ArticleVersionRow | null> {
    const [version] = await this.db
      .select()
      .from(articleVersions)
      .where(
        and(
          eq(articleVersions.articleId, articleId),
          eq(articleVersions.id, versionId),
        ),
      )
      .limit(1)
    return version ?? null
  }

  async getVersionForWorkspace(
    articleId: string,
    versionId: number,
    workspaceId: string,
  ): Promise<ArticleVersionRow | null> {
    if (!(await this.getArticleForWorkspace(articleId, workspaceId))) return null
    return this.getVersion(articleId, versionId)
  }

  async patchArticle(input: ArticleWriteInput): Promise<ArticleMutationResult> {
    const content = normalizeContent(input.content)
    return this.mutateCurrent(input.articleId, input.expectedRevision, async () => ({
      content,
      contentFingerprint: contentFingerprint(content),
      wordCount: wordCount(content),
    }))
  }

  async patchArticleForWorkspace(
    input: ArticleWriteInput,
    workspaceId: string,
  ): Promise<ArticleMutationResult> {
    const content = normalizeContent(input.content)
    return this.mutateCurrent(input.articleId, input.expectedRevision, async () => ({
      content,
      contentFingerprint: contentFingerprint(content),
      wordCount: wordCount(content),
    }), workspaceId)
  }

  async restoreVersion(input: RestoreArticleVersionInput): Promise<ArticleMutationResult> {
    return this.mutateCurrent(
      input.articleId,
      input.expectedRevision,
      async (tx) => {
        const [version] = await tx
          .select()
          .from(articleVersions)
          .where(
            and(
              eq(articleVersions.articleId, input.articleId),
              eq(articleVersions.id, input.versionId),
            ),
          )
          .limit(1)
        if (!version) return null
        return {
          content: version.content,
          contentFingerprint: version.contentFingerprint,
          wordCount: version.wordCount,
        }
      },
    )
  }

  async restoreVersionForWorkspace(
    input: RestoreArticleVersionInput,
    workspaceId: string,
  ): Promise<ArticleMutationResult> {
    return this.mutateCurrent(
      input.articleId,
      input.expectedRevision,
      async (tx) => {
        const [version] = await tx
          .select()
          .from(articleVersions)
          .where(
            and(
              eq(articleVersions.articleId, input.articleId),
              eq(articleVersions.id, input.versionId),
            ),
          )
          .limit(1)
        if (!version) return null
        return {
          content: version.content,
          contentFingerprint: version.contentFingerprint,
          wordCount: version.wordCount,
        }
      },
      workspaceId,
    )
  }

  private async mutateCurrent(
    articleId: string,
    expectedRevision: number,
    nextContent: (
      tx: Parameters<Parameters<VibeDatabase<TQueryResult>['transaction']>[0]>[0],
      current: ArticleRow,
    ) => Promise<{
      content: string
      contentFingerprint: string
      wordCount: number
    } | null>,
    workspaceId?: string,
  ): Promise<ArticleMutationResult> {
    return this.db.transaction(async (tx) => {
      const [current] = await tx
        .select()
        .from(articles)
        .where(
          workspaceId
            ? and(
                eq(articles.id, articleId),
                sql`exists (
                  select 1 from ${jobs}
                  where ${jobs.id} = ${articles.jobId}
                    and ${jobs.workspaceId} = ${workspaceId}
                )`,
              )
            : eq(articles.id, articleId),
        )
        .for('update')
        .limit(1)
      if (!current) return { status: 'not_found' as const }
      if (current.revision !== expectedRevision) {
        return {
          status: 'revision_conflict' as const,
          currentRevision: current.revision,
        }
      }

      const replacement = await nextContent(tx, current)
      if (!replacement) return { status: 'version_not_found' as const }

      const [snapshot] = await tx
        .insert(articleVersions)
        .values({
          articleId: current.id,
          sourceRevision: current.revision,
          content: current.content,
          contentFingerprint: current.contentFingerprint,
          wordCount: current.wordCount,
        })
        .returning()
      if (!snapshot) throw new Error(`Article snapshot failed for ${articleId}`)

      const [article] = await tx
        .update(articles)
        .set({
          ...replacement,
          revision: current.revision + 1,
          updatedAt: sql`clock_timestamp()`,
        })
        .where(
          and(
            eq(articles.id, articleId),
            eq(articles.revision, expectedRevision),
          ),
        )
        .returning()
      if (!article) throw new Error(`Article revision changed for ${articleId}`)
      return { status: 'updated' as const, article, snapshot }
    })
  }
}

export function createArticleRepository<TQueryResult extends PgQueryResultHKT>(
  db: VibeDatabase<TQueryResult>,
) {
  return new ArticleRepository(db)
}
